agent/consul/controller/controller.go

```diff
@@ -0,0 +1,258 @@
+type controller struct {
+	// reconciler is the Reconciler that processes all subscribed
+	// Requests
+	reconciler Reconciler
+
+	// makeQueue is the factory used for creating the work queue, generally
+	// this shouldn't be touched, but can be updated for testing purposes
+	makeQueue func(ctx context.Context, baseBackoff time.Duration, maxBackoff time.Duration) WorkQueue
+	// workers is the number of workers to use to process data
+	workers int
+	// work is the internal work queue that pending Requests are added to
+	work WorkQueue
+	// baseBackoff is the starting backoff time for the work queue's rate limiter
+	baseBackoff time.Duration
+	// maxBackoff is the maximum backoff time for the work queue's rate limiter
+	maxBackoff time.Duration
+
+	// subscriptions is a list of subscription requests for retrieving configuration entries
+	subscriptions []subscription
+	// publisher is the event publisher that should be subscribed to for any updates
+	publisher state.EventPublisher
+}
+
+// New returns a new Controller associated with the given state store and reconciler.
+func New(publisher state.EventPublisher, reconciler Reconciler) Controller {
+	return &controller{
+		reconciler:  reconciler,
+		publisher:   publisher,
+		workers:     1,
+		baseBackoff: 5 * time.Millisecond,
+		maxBackoff:  1000 * time.Second,
+		makeQueue:   RunWorkQueue,
+	}
+}
+
+// Subscribe tells the controller to subscribe to updates for config entries of the
+// given kind and with the associated enterprise metadata. This should only ever be
+// called prior to running Start.
+func (c *controller) Subscribe(request *stream.SubscribeRequest, transformers ...func(entry structs.ConfigEntry) []Request) Controller {
+	c.subscriptions = append(c.subscriptions, subscription{
+		request:      request,
+		transformers: transformers,
+	})
+	return c
+}
+
+// WithBackoff changes the base and maximum backoff values for the Controller's
+// Request retry rate limiter. This should only ever be called prior to
+// running Start.
+func (c *controller) WithBackoff(base, max time.Duration) Controller {
+	c.baseBackoff = base
+	c.maxBackoff = max
+	return c
+}
+
+// WithWorkers sets the number of worker goroutines used to process the queue
+// this defaults to 1 goroutine.
+func (c *controller) WithWorkers(i int) Controller {
+	if i <= 0 {
+		i = 1
+	}
+	c.workers = i
+	return c
+}
+
+// WithQueueFactory changes the initialization method for the Controller's work
+// queue, this is predominantly just used for testing. This should only ever be called
+// prior to running Start.
+func (c *controller) WithQueueFactory(fn func(ctx context.Context, baseBackoff time.Duration, maxBackoff time.Duration) WorkQueue) Controller {
+	c.makeQueue = fn
+	return c
+}
+
+// Start begins the Controller's main processing loop. When the given
+// context is canceled, the Controller stops processing any remaining work.
+// The Start function should only ever be called once.
```

Comment:
Would use `sync.Once` inside `Start( ... )` so that we don't rely on the caller to ensure they're only calling once

Reply:
👍 Either approach is an improvement IMO

Reply:
Cool, added some panic-based assertions (with tests) around `Run` and the configuration functions to enforce this requirement in the concrete implementation.